cpp/src/link_prediction/similarity_impl.cuh

```diff
@@ -162,5 +163,302 @@ rmm::device_uvector<weight_t> similarity(
+        edge_t next_boundary;
+        raft::update_host(
+          &next_boundary, two_hop_degrees.data() + current_pos, 1, handle.get_stream());
+        next_boundary += MAX_PAIRS;
+
+        next_pos =
+          current_pos + thrust::distance(two_hop_degrees.begin() + current_pos,
+                                         thrust::upper_bound(handle.get_thrust_policy(),
+                                                             two_hop_degrees.begin() + current_pos,
+                                                             two_hop_degrees.end(),
+                                                             next_boundary));
+
+        if (next_pos == current_pos) next_pos++;
+      }
+
+      size_t batch_size = next_pos - current_pos;
+
+      if constexpr (multi_gpu) {
+        batch_size = cugraph::host_scalar_allreduce(
+          handle.get_comms(), batch_size, raft::comms::op_t::SUM, handle.get_stream());
+      }
+
+      if (batch_size == 0) break;
+
+      auto [offsets, v2] = k_hop_nbrs(
+        handle,
+        graph_view,
+        raft::device_span<vertex_t const>{sources.data() + current_pos, next_pos - current_pos},
+        2,
+        do_expensive_check);
+
+      auto v1 = cugraph::detail::expand_sparse_offsets(
+        raft::device_span<size_t const>{offsets.data(), offsets.size()},
+        vertex_t{0},
+        handle.get_stream());
+
+      cugraph::unrenumber_local_int_vertices(handle,
+                                             v1.data(),
+                                             v1.size(),
+                                             sources.data() + current_pos,
+                                             vertex_t{0},
+                                             static_cast<vertex_t>(next_pos - current_pos),
+                                             do_expensive_check);
+
+      auto new_size = thrust::distance(
+        thrust::make_zip_iterator(v1.begin(), v2.begin()),
+        thrust::remove_if(
+          handle.get_thrust_policy(),
+          thrust::make_zip_iterator(v1.begin(), v2.begin()),
+          thrust::make_zip_iterator(v1.end(), v2.end()),
+          [] __device__(auto tuple) { return thrust::get<0>(tuple) == thrust::get<1>(tuple); }));
+
+      v1.resize(new_size, handle.get_stream());
+      v2.resize(new_size, handle.get_stream());
+
+      auto score =
+        similarity(handle,
+                   graph_view,
+                   edge_weight_view,
+                   std::make_tuple(raft::device_span<vertex_t const>{v1.data(), v1.size()},
+                                   raft::device_span<vertex_t const>{v2.data(), v2.size()}),
+                   functor,
+                   do_expensive_check);
+
+      thrust::sort_by_key(handle.get_thrust_policy(),
+                          score.begin(),
+                          score.end(),
+                          thrust::make_zip_iterator(v1.begin(), v2.begin()),
+                          thrust::greater<weight_t>{});
```

Comment:
In case top_v1.size() >= topk (in case of multi-gpu, aggregate top_v1.size() >= topk), we know the minimum threshold to be included in topk, and we can run thrust::remove_if first to avoid sorting a large array.